### ISSUE TITLE
Questions about the USB Demod and Data format going into JAERO

### ISSUE THREAD

**Issue by muaddib1984** [jeroenbeijer/SDRReceiver#18]
Hi there. I've been trying to emulate your receiver in GNURadio. I'm pretty experienced with GNURadio and my primary motivation is to enable use of wideband SDRs with JAERO. I've been discussing and sharing my progress with Jontio [here](https://github.com/jontio/JAERO/issues/67).  
Thus far, i've been testing my flowgraphs with audio WAV and I/Q WAV data from files provided by @jontio and now that i've got my SDR L-Band setup working i'm taking 20MHz I/Q captures with my USRP and trying to get those to work, that's where i'm stuck. 
While I'm successfully decoding a few frames from 1 or 2 of the 600 bps channels, the majority of the data frames are 'Bad CRC'. My capture has decent SNR (6-10dB i think). I'm handling all the channelization (VFO) and decimation to get the signal down to 48ksps, then implementing a USB receiver as best I can, then I have a custom block that slightly changes the ZMQ structure to send appropriate rate messages and 16-bit shorts to JAERO.

My question is related to your USB demod and the data type you're sending into JAERO. I've looked through vfo.cpp pretty thoroughly.

My questions are:
1) Does SDRReceiver ever send the I/Q to JAERO or is it always real-valued float data? 
I see that you have the demodUSB boolean [here](https://github.com/jeroenbeijer/SDRReceiver/blob/700bdfa6a5bd517f9a6b8e67cca5342d06622954/vfo.cpp#L426-L453) that sends real valued float data, but the `else if` right after that sends I/Q if there's a 'topic' (presumably 'vfo00'). Does the code ever reach that point?
2) I've also been trying to mimic your USB demod by focusing mainly on these:
https://github.com/jeroenbeijer/SDRReceiver/blob/700bdfa6a5bd517f9a6b8e67cca5342d06622954/vfo.cpp#L309
https://github.com/jeroenbeijer/SDRReceiver/blob/700bdfa6a5bd517f9a6b8e67cca5342d06622954/vfo.cpp#L318
It looks like you are starting with complex decimated data from the VFO channel and multiplying that complex data by the beat frequency oscillator (the data rate of the channel?), then applying low-pass FIR and subtracting the the hilbert transformed imaginary part from the low-passed real part. Is that all correct? 

You've really setup a cool thing here and I really like that you've used ZMQ to get away from the audio discrepancies.

Thanks!

**Comment by jeroenbeijer** [comment]
Hi!

Sounds like an interesting project! 

There is some code in there that I used for a different purpose, i.e. feeding a part of the bandwidth from one SDR to another "IQ" receiver somewhere else. It was over a high latency link so I dropped some bits of each sample to reduce bandwidth. But this is not used in the current setup unless you add some more stuff in the ini file. But no, currently it sends USB 16 bit shorts to jaero. 

The osc_bfo is not really used any longer, sorry, I should clean this up but you know, time is short. Just looked at the code again to refresh my own memory. Basically with a typical ini file (i.e. sample rate 1536000) the usb demod takes place in usb_demod which cycles through the IQ data in the final decimation vector. For each IQ pair it sends one arm through a delay line and the other through a hilbert transform to get the 90 degree rotation. The delay line is to line them back up when they come out. Subtract imag from real and you get USB:

usb = delayT.update_dont_touch(curr.real()) - philbert->FIRUpdateAndProcess(curr.imag());

if the ini file tells it to also low pass the output audio it will do this:

usb = fir_usb->FIRUpdateAndProcess(delayT.update_dont_touch(curr.real()) - philbert->FIRUpdateAndProcess(curr.imag()));

So basically just to FIR the USB data. It then converts the USB floats to 16 bit shorts:

transmit_usb[i] = usb * gain * 32768.0;

I guess one thing to keep in mind, for 600 bps the standard sample rate is 12Khz out of the SDR into Jaero. But jaero also accepts 48Khz provided of course you send the rate appropriately in the ZMQ rate messages that come before each data buffer.

I am a total novice at most of this DSP stuff but I recall I learned a lot from this tutorial (and the other ones on that site)

https://www.g0kla.com/sdr/tutorials/sdr_tutorial7.php 

BR

Jeroen

**Comment by muaddib1984** [comment]
@jeroenbeijer thank you so much! This clears up a lot. The tutorial that you linked has helped me sanity check my usb demodulator. I applied some small changes based on your guidance and now can get at least one 600bps channel to decode in real-time with very few 'Bad CRC' messages. I also think that this may be related to good old SNR. I have a new antenna on the way so I hope to be further ruling out front end issues soon. 

Thanks again for spending the time to answer me. I agree, time is a precious thing for us all!

**Comment by muaddib1984** [comment]
@jeroenbeijer A quick update. I'm now able to USB demod and decode 600 and 1200 bps! I seem to be having some trouble with 8400 and 10500. Did you have many difficulties demodding those signals compared to 600/1200?  I know they tend to be quieter. 
Wanted to show you the results of 600/1200 anyways.
![JAERO_1200bps](https://user-images.githubusercontent.com/70996692/154406011-2628f8ca-0d98-48c8-8948-9acece85d57a.png)
![JAERO_600_1200_spectrum](https://user-images.githubusercontent.com/70996692/154406026-ead1be2c-44d5-4a69-8623-5baf624d3d29.png)
![JAERO_USB_demod](https://user-images.githubusercontent.com/70996692/154406132-65ed4fd8-7ebb-46dd-aed2-edba29117bc9.png)

**Comment by muaddib1984** [comment]
Got everything working! Thank you so much for your help!
Here's my [project](https://github.com/muaddib1984/gr-JAERO).

**Comment by muaddib1984** [comment]
I've been able to decode 600,1200,8400,10500 and C-Band 10500 burst messages, closing the issue
